# SQL - выпускной проект

Выполнил Паньков.О

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загружаем-библиотеки" data-toc-modified-id="Загружаем-библиотеки-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загружаем библиотеки</a></span></li><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Ознакомление-с-таблицами" data-toc-modified-id="Ознакомление-с-таблицами-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Ознакомление с таблицами</a></span></li><li><span><a href="#Посчитать,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитать,-сколько-книг-вышло-после-1-января-2000-года-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Посчитать, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитать-количество-обзоров-и-среднюю-оценку-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Для каждой книги посчитать количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц" data-toc-modified-id="Определить-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Определить издательство, которое выпустило наибольшее число книг толще 50 страниц</a></span></li><li><span><a href="#Определить-автора-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определить-автора-с-самой-высокой-средней-оценкой-книг-—-учитывать-только-книги-с-50-и-более-оценками-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок" data-toc-modified-id="Посчитать-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-48-оценок-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок</a></span></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Проект по анализу книжной базы данных для создания приложения

Цель исследования: найти информацию о самых востребованных книжных предпочтениях читателей для формирования ценностного предложения нового продукта.

Задачи исследования:

* посчитать, сколько книг вышло после 1 января 2000 года;
* для каждой книги посчитать количество обзоров и среднюю оценку;
* определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
* определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
* посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# Описание данных

Таблица books Содержит данные о книгах:

* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

Таблица authors Содержит данные об авторах:

* author_id — идентификатор автора;
* author — имя автора.

Таблица publishers Содержит данные об издательствах:

* publisher_id — идентификатор издательства;
* publisher — название издательства;

Таблица ratings Содержит данные о пользовательских оценках книг:

* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

Таблица reviews Содержит данные о пользовательских обзорах:

* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя автора обзора;
* text — текст обзора.

## Загружаем библиотеки 

In [1]:
import pandas as pd
import numpy as np

from sqlalchemy import create_engine

pd.set_option('display.max_colwidth', None)

## Подключение к базе данных

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


## Ознакомление с таблицами

In [3]:
display(pd.io.sql.read_sql('SELECT * FROM books LIMIT 5', con=engine) )
display(pd.io.sql.read_sql('SELECT * FROM authors LIMIT 5', con=engine) )
display(pd.io.sql.read_sql('SELECT * FROM publishers LIMIT 5', con=engine) )
display(pd.io.sql.read_sql('SELECT * FROM ratings LIMIT 5', con=engine) )
display(pd.io.sql.read_sql('SELECT * FROM reviews LIMIT 5', con=engine) )

book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

Создадим функцию для вывода результата

In [4]:
def output(sql):
    return pd.read_sql(sql,con = engine)

## Посчитать, сколько книг вышло после 1 января 2000 года

In [5]:
count_books = """ SELECT COUNT(book_id)
                  FROM books
                  WHERE publication_date > CAST('2000-01-01' as date)
              """
output(count_books)

count
0    819

После 1 января 2000 года вышло 819 книг

## Для каждой книги посчитать количество обзоров и среднюю оценку

In [6]:
count_reviews_and_means = """ WITH
                              rev AS
                              (SELECT book_id, 
                                      COUNT(review_id) OVER (PARTITION BY book_id) AS count
                              FROM reviews),
                              rat AS 
                              (SELECT book_id, 
                                      ROUND(AVG(rating) OVER (PARTITION BY book_id), 2) AS avg_rating
                              FROM ratings) 

                              SELECT DISTINCT(b.book_id),
                                     b.title, 
                                     count, 
                                     avg_rating
                              FROM books AS b
                              LEFT JOIN rev ON rev.book_id = b.book_id
                              LEFT JOIN rat ON rat.book_id = b.book_id
                              ORDER BY 1 
                              
                          """
output(count_reviews_and_means)

book_id  \
0          1   
1          2   
2          3   
3          4   
4          5   
..       ...   
995      996   
996      997   
997      998   
998      999   
999     1000   

                                                                                title  \
0                                                                        'Salem's Lot   
1                                                  1 000 Places to See Before You Die   
2                                 13 Little Blue Envelopes (Little Blue Envelope  #1)   
3                               1491: New Revelations of the Americas Before Columbus   
4                                                                                1776   
..                                                                                ...   
995                                          Wyrd Sisters (Discworld  #6; Witches #2)   
996                                                       Xenocide (Ender's Saga  #3)   
997                                                                   Year of Wonders   
998                                                       You Suck (A Love Story  #2)   
999  Zen and the Art of Motorcycle Maintenance: An Inquiry Into Values (Phaedrus  #1)   

     count  avg_rating  
0      2.0        3.67  
1      1.0        2.50  
2      3.0        4.67  
3      2.0        4.50  
4      4.0        4.00  
..     ...         ...  
995    3.0        3.67  
996    3.0        3.40  
997    4.0        3.20  
998    2.0        4.50  
999    4.0        3.83  

[1000 rows x 4 columns]

Посчитано количество обзоров и средняя оценка для каждой книги

## Определить издательство, которое выпустило наибольшее число книг толще 50 страниц

In [7]:
publisher = """ SELECT publishers.publisher,
                      COUNT(books.book_id) AS count
                FROM publishers
                INNER JOIN books ON books.publisher_id = publishers.publisher_id
                WHERE books.num_pages > 50
                GROUP BY 1
                ORDER BY 2 DESC
                LIMIT 10
            """

output(publisher)

publisher  count
0              Penguin Books     42
1                    Vintage     31
2   Grand Central Publishing     25
3           Penguin Classics     24
4           Ballantine Books     19
5                     Bantam     19
6                    Berkley     17
7         St. Martin's Press     14
8              Berkley Books     14
9  William Morrow Paperbacks     13

Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц

## Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками

In [16]:
author_max = """ WITH rat AS (SELECT book_id, 
                                    COUNT(rating_id) 
                             FROM ratings
                             GROUP BY 1
                             HAVING COUNT(rating_id) > 50)
            
                SELECT a.author,
                       AVG(r.rating) 
                FROM books AS b
                INNER JOIN ratings AS r ON r.book_id = b.book_id 
                INNER JOIN authors AS a ON a.author_id = b.author_id
                WHERE b.book_id IN (SELECT book_id FROM rat)
                GROUP BY 1
                ORDER BY 2 DESC
                
                LIMIT 1
            """

output(author_max)



author       avg
0  J.K. Rowling/Mary GrandPré  4.287097

Автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками - J.K. Rowling

## Посчитать среднее количество обзоров от пользователей, которые поставили больше 48 оценок

In [9]:
review_avg = """ SELECT AVG(review.review_cnt) AS avg_review
                          FROM (SELECT COUNT(review_id) AS review_cnt
                                FROM reviews
                                WHERE username IN (SELECT username
                                                   FROM ratings
                                                   GROUP BY 1
                                                   HAVING COUNT(ratings) > 48)
                               GROUP BY username) AS review; 
             """
output(review_avg)

avg_review
0        24.0

24 - среднее количество обзоров от пользователей, которые поставили больше 48 оценок

## Выводы

* После 1 января 2000 года вышло 819 книг
* Посчитано количество обзоров и средняя оценка для каждой книги
* Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц
* Автор с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками - J.K. Rowling
* 24 - среднее количество обзоров от пользователей, которые поставили больше 48 оценок